### PR TITLE
Adding the payload-accepted and payload-rejected controllers

### PR DESCRIPTION
**Pull request by bradmwilliams**

### CODE REVIEW

File: pkg/apis/release/v1alpha1/types.go [hunk 1]
```diff
@@ -204,6 +204,27 @@ const (
 	ConditionPayloadRejected string = "PayloadRejected"
 )
 
+// These are valid Reason values for the ReleasePayloadStatus Conditions
```
Comment (deads2k) [inline pkg/apis/release/v1alpha1/types.go:207]: These aren't API, move them to your controller.

File: pkg/cmd/release-payload-controller/payload_accepted_controller.go [hunk 2]
```diff
@@ -3,10 +3,16 @@ package release_payload_controller
...
+	"github.com/openshift/release-controller/pkg/apis/release/v1alpha1"
 	releasepayloadclient "github.com/openshift/release-controller/pkg/client/clientset/versioned/typed/release/v1alpha1"
 	releasepayloadinformer "github.com/openshift/release-controller/pkg/client/informers/externalversions/release/v1alpha1"
 	releasepayloadlister "github.com/openshift/release-controller/pkg/client/listers/release/v1alpha1"
+	"github.com/openshift/release-controller/pkg/releasepayload/conditions"
+	"github.com/openshift/release-controller/pkg/releasepayload/jobstatus"
+	releasepayloadhelpers "github.com/openshift/release-controller/pkg/releasepayload/v1alpha1helpers"
 	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"reflect"
```
Comment (deads2k) [inline pkg/cmd/release-payload-controller/payload_accepted_controller.go:15]: missing the writes and reads doc on the controller.

File: pkg/cmd/release-payload-controller/payload_accepted_controller.go [hunk 3]
```diff
@@ -133,9 +139,75 @@ func (c *PayloadAcceptedController) sync(ctx context.Context, key string) error
 		return err
 	}
 
+	originalCondition := conditions.GetReleasePayloadCondition(originalReleasePayload, v1alpha1.ConditionPayloadAccepted)
```
Comment (deads2k) [inline pkg/cmd/release-payload-controller/payload_accepted_controller.go:142]: make a function in library-go that takes []Conditions and returns you the *Condition

File: pkg/cmd/release-payload-controller/payload_accepted_controller.go [hunk 4]
```diff
@@ -133,9 +139,75 @@ func (c *PayloadAcceptedController) sync(ctx context.Context, key string) error
...
+		return nil
+	}
+	if err != nil {
+		return err
+	}
 
 	return nil
 }
+
+func computeReleasePayloadAcceptedCondition(payload *v1alpha1.ReleasePayload, originalCondition *metav1.Condition) (*metav1.Condition, bool) {
```
Comment (deads2k) [inline pkg/cmd/release-payload-controller/payload_accepted_controller.go:174]: godoc to tell me what this bool means.

File: pkg/cmd/release-payload-controller/payload_accepted_controller.go [hunk 5]
```diff
@@ -133,9 +139,75 @@ func (c *PayloadAcceptedController) sync(ctx context.Context, key string) error
...
+	acceptedCondition := &metav1.Condition{
+		Type:               v1alpha1.ConditionPayloadAccepted,
+		Status:             metav1.ConditionUnknown,
+		Reason:             v1alpha1.ReleasePayloadAcceptedReason,
+		LastTransitionTime: metav1.Now(),
```
Comment (deads2k) [inline pkg/cmd/release-payload-controller/payload_accepted_controller.go:179]: no

Comment (deads2k) [reply pkg/cmd/release-payload-controller/payload_accepted_controller.go:179]: You want logic like this: https://github.com/openshift/library-go/blob/master/pkg/operator/v1helpers/helpers.go#L54-L72

File: pkg/releasepayload/jobstatus/helpers.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+		default:
+			unknownJobs = append(unknownJobs, job)
+		}
+	}
+
+	switch {
+	// If there are any pending jobs, then we're still working on validating the release
+	case len(pendingJobs) > 0:
+		return v1alpha1.JobStatePending
+	// If there are any unknown jobs, then we won't be able to figure out a valid state
```
Comment (deads2k) [inline pkg/releasepayload/jobstatus/helpers.go:43]: This isn't true.  If you have a failure, you failed.

File: pkg/releasepayload/jobstatus/helpers.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+	for _, job := range jobs {
+		switch job.AggregateState {
+		case v1alpha1.JobStatePending:
+			pendingJobs = append(pendingJobs, job)
+		case v1alpha1.JobStateSuccess:
+			successfulJobs = append(successfulJobs, job)
+		case v1alpha1.JobStateFailure:
+			failedJobs = append(failedJobs, job)
+		case v1alpha1.JobStateIgnored:
```
Comment (deads2k) [inline pkg/releasepayload/jobstatus/helpers.go:32]: This sounds like they should be informing and remove the JobStateIgnored.

File: pkg/cmd/release-payload-controller/payload_accepted_controller.go [hunk 8]
```diff
@@ -133,9 +139,75 @@ func (c *PayloadAcceptedController) sync(ctx context.Context, key string) error
...
+			return acceptedCondition, true
+		}
+		// If a payload has already been rejected, then we do not want to process it any further...
+		return nil, false
+	}
+
+	// Check that all Blocking jobs have completed successfully...
+	status := jobstatus.ComputeAggregatedJobState(payload.Status.BlockingJobResults)
+	switch status {
+	case v1alpha1.JobStateSuccess, v1alpha1.JobStateIgnored:
```
Comment (deads2k) [inline pkg/cmd/release-payload-controller/payload_accepted_controller.go:204]: so no ignored here either.

File: pkg/cmd/release-payload-controller/payload_accepted_controller.go [hunk 9]
```diff
@@ -133,9 +139,75 @@ func (c *PayloadAcceptedController) sync(ctx context.Context, key string) error
...
+	case v1alpha1.ReleasePayloadOverrideRejected:
+		// Update previously Accepted condition
+		if originalCondition != nil {
+			acceptedCondition.Status = metav1.ConditionFalse
+			acceptedCondition.Message = originalCondition.Message
+			acceptedCondition.Reason = originalCondition.Reason
+			return acceptedCondition, true
+		}
+		// If a payload has already been rejected, then we do not want to process it any further...
+		return nil, false
```
Comment (deads2k) [inline pkg/cmd/release-payload-controller/payload_accepted_controller.go:198]: this should unconditionally return the Accepted condition as False.

File: pkg/cmd/release-payload-controller/payload_accepted_controller.go (hunk 4 above)
Comment (deads2k) [reply pkg/cmd/release-payload-controller/payload_accepted_controller.go:174]: meaning this has no bool and is not a pointer.

File: pkg/cmd/release-payload-controller/payload_accepted_controller.go [hunk 10]
```diff
@@ -133,9 +139,75 @@ func (c *PayloadAcceptedController) sync(ctx context.Context, key string) error
 		return err
 	}
 
+	originalCondition := conditions.GetReleasePayloadCondition(originalReleasePayload, v1alpha1.ConditionPayloadAccepted)
+	acceptedCondition, ok := computeReleasePayloadAcceptedCondition(originalReleasePayload, originalCondition)
+	if !ok {
+		return nil
+	}
+
+	if !conditions.ConditionChanged(originalCondition, acceptedCondition) {
```
Comment (deads2k) [inline pkg/cmd/release-payload-controller/payload_accepted_controller.go:148]: you want the library-go SetoperatorCondition like logic.

File: pkg/apis/release/v1alpha1/types.go (hunk 1 above)
Comment (bradmwilliams) [reply pkg/apis/release/v1alpha1/types.go]: Done!

File: pkg/cmd/release-payload-controller/payload_accepted_controller.go (hunk 2 above)
Comment (bradmwilliams) [reply pkg/cmd/release-payload-controller/payload_accepted_controller.go:15]: Done!

File: pkg/cmd/release-payload-controller/payload_accepted_controller.go (hunk 4 above)
Comment (bradmwilliams) [reply pkg/cmd/release-payload-controller/payload_accepted_controller.go]: Done!

File: pkg/cmd/release-payload-controller/payload_accepted_controller.go (hunk 5 above)
Comment (bradmwilliams) [reply pkg/cmd/release-payload-controller/payload_accepted_controller.go]: Ack.  Done!

File: pkg/cmd/release-payload-controller/payload_accepted_controller.go (hunk 8 above)
Comment (bradmwilliams) [reply pkg/cmd/release-payload-controller/payload_accepted_controller.go]: Done!

File: pkg/cmd/release-payload-controller/payload_accepted_controller.go (hunk 9 above)
Comment (bradmwilliams) [reply pkg/cmd/release-payload-controller/payload_accepted_controller.go]: Done!

File: pkg/cmd/release-payload-controller/payload_accepted_controller.go (hunk 10 above)
Comment (bradmwilliams) [reply pkg/cmd/release-payload-controller/payload_accepted_controller.go]: Done!

File: pkg/cmd/release-payload-controller/payload_rejected_controller.go [hunk 11]
```diff
@@ -133,9 +157,64 @@ func (c *PayloadRejectedController) sync(ctx context.Context, key string) error
...
+	switch payload.Spec.PayloadOverride.Override {
+	case v1alpha1.ReleasePayloadOverrideRejected:
+		rejectedCondition.Status = metav1.ConditionTrue
+		rejectedCondition.Message = payload.Spec.PayloadOverride.Reason
+		rejectedCondition.Reason = ReleasePayloadManuallyRejectedReason
+		return rejectedCondition
+	case v1alpha1.ReleasePayloadOverrideAccepted:
+		rejectedCondition.Status = metav1.ConditionFalse
+		// Update previously Rejected condition
+		if originalCondition != nil {
```
Comment (deads2k) [inline pkg/cmd/release-payload-controller/payload_rejected_controller.go:200]: same comment here, don't use the old condition.

Comment (bradmwilliams) [reply pkg/cmd/release-payload-controller/payload_rejected_controller.go]: Done!

File: pkg/cmd/release-payload-controller/payload_accepted_controller.go [hunk 12]
```diff
@@ -133,9 +157,58 @@ func (c *PayloadAcceptedController) sync(ctx context.Context, key string) error
...
+	// Check for "Accepted" PayloadOverride
+	switch payload.Spec.PayloadOverride.Override {
+	case v1alpha1.ReleasePayloadOverrideAccepted:
+		acceptedCondition.Status = metav1.ConditionTrue
+		acceptedCondition.Message = payload.Spec.PayloadOverride.Reason
+		acceptedCondition.Reason = ReleasePayloadManuallyAcceptedReason
+		return acceptedCondition
+	case v1alpha1.ReleasePayloadOverrideRejected:
+		// If a payload has already been rejected, then we do not want to process it any further...
+		acceptedCondition.Status = metav1.ConditionFalse
```
Comment (deads2k) [inline pkg/cmd/release-payload-controller/payload_accepted_controller.go:198]: you forgot to set reason and message

Comment (bradmwilliams) [reply pkg/cmd/release-payload-controller/payload_accepted_controller.go:198]: Reason is set above L186.  Is Message required if I'm just blanking the field?

Comment (bradmwilliams) [reply pkg/cmd/release-payload-controller/payload_accepted_controller.go:198]: Done!

File: pkg/releasepayload/conditions/helpers.go [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+func (in ByReleasePayloadConditionType) Swap(i, j int) {
+	in[i], in[j] = in[j], in[i]
+}
+
+// The following logic is straight from: https://github.com/openshift/library-go/pull/1341
+// I'm just copying it here to prevent having to do the vendoring dance and we can pick
+// up the new version of library-go when the PR merges.
+
+func SetCondition(conditions *[]metav1.Condition, newCondition metav1.Condition) {
```
Comment (deads2k) [inline pkg/releasepayload/conditions/helpers.go:27]: followup: remove in favor of https://github.com/openshift/library-go/pull/1341 please

**Pull request merged**